flybywiresim/a32nx
# feat(fms): cFMS 1.5

Credit @professoralex13 @tracernz for the help.

## Summary of Changes

* This PR introduces support for the following FMS features
  * LNAV
    * Holding patterns
    * Turn direction constraints on non-TF legs
    * Overfly restriction support
    * ARINC424 leg types
      * AF, CA, CI, CR, CF, DF, HF, HM, HA legs
    * Turn prediction types
      * Path capture
      * Course capture
      * Direct to fix turn
      * Holding pattern entry turn   
  * ND
    * Removed flight plan loading from `localStorage`
    * Corrected active waypoint ETA display
    * Added `EfisVectors` systems with optimized transmit task queue + support for **future** display of OFFSET, SECONDARY, SECONDARY DASHED, MISSED APPROACH, ALTERNATE and EOSID flight paths.

## Known limitations

- CA leg terminations are sometimes in the wrong place, and do not adapt to V/S.
- Some path captures will be incorrectly drawn. This will not affect guidance.
- INTCPT calculation can be off on large distances.
- Course captures do not adapt to PPOS when nextLeg is active. This can cause the final path to be off to the side.

## Screenshots (if necessary)

**LSZH DEGES3W Before**

![image](https://user-images.githubusercontent.com/4503241/145327972-79b57b90-c874-4f1c-9399-e25f293e4ccc.png)

**LSZH DEGES3W After**

![image](https://user-images.githubusercontent.com/4503241/145327987-249f3b25-9a6b-45ec-9ccf-ab62a58ea562.png)

**Overfly support on TF legs**

![image](https://user-images.githubusercontent.com/4503241/145328060-7aebe5f9-4c06-4e48-87c3-4ce3d025ec84.png)

**Direct To Fix turns**

![image](https://user-images.githubusercontent.com/4503241/145328072-5c2ba622-4bde-4ab4-a974-1fdef6e674e2.png)

**Honeywell path captures**

![image](https://user-images.githubusercontent.com/4503241/145328104-eed9e7be-b2f5-48e0-8e4e-506f12f34746.png)

## References

An absolutely enormous amount of them and a year+ of conception.




## Additional context



Discord username (if different from GitHub): someperson#4953

## Testing instructions


PHASE 1 - General LNAV and path generation

* Make sure to test loading various SIDs and STARs and that their drawing is reasonable
  * Test departures with turn direction constraints
* Make sure to test flying various SIDs and STARs and that the aircraft flies it correctly
  * Keep in mind that path capture turns (sharp lines joining the next leg with a 45 degree angle) will not be followed exactly - this is accurate to the IRL Honeywell FMS 
  * Check if turns change shape / size while flying them (they should not)
* Make sure sequencing and guidance work correctly (no turning the wrong way, minimal overshooting / undershooting)
  * Make sure VM (MANUAL) legs do not sequence and do not show a distance or ETA on the ND but do show a bearing
  * Make sure regular legs show a distance, bearing and ETA on the NDD
* Make sure ND does not crash and that there is no unacceptable delay in path drawing 

PHASE 2 - Holding patterns, entries and exits (not available yet)

TBD


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
